src/anyp/Uri.cc

```diff
@@ -438,24 +421,52 @@ AnyP::Uri::parse(const HttpRequestMethod& method, const char *url)
         }
     }
 
-    parseFinish(protocol, proto, urlpath, foundHost, SBuf(login), foundPort);
+    setScheme(scheme);
+    path(urlpath);
+    host(foundHost);
+    userInfo(SBuf(login));
+    port(foundPort);
     return true;
 }
 
-/// Update the URL object with parsed URI data.
-void
-AnyP::Uri::parseFinish(const AnyP::ProtocolType protocol,
-                       const char *const protoStr, // for unknown protocols
-                       const char *const aUrlPath,
-                       const char *const aHost,
-                       const SBuf &aLogin,
-                       const int aPort)
+/**
+ * Governerd by RFC 8141 section 2:
+ *
+ *  assigned-name = "urn" ":" NID ":" NSS
+ *  NID           = (alphanum) 0*30(ldh) (alphanum)
+ *  ldh           = alphanum / "-"
+ *  NSS           = pchar *(pchar / "/")
+ *
+ * Notice that NID is exactly 2-32 characters in length.
+ */
+bool
+AnyP::Uri::parseUrn(Parser::Tokenizer &tok)
 {
-    setScheme(protocol, protoStr);
-    path(aUrlPath);
-    host(aHost);
-    userInfo(aLogin);
-    port(aPort);
+    static const CharacterSet nidChars = CharacterSet("NID","-") + CharacterSet::ALPHANUM;
+    SBuf nid;
+    if (tok.prefix(nid, nidChars, 32) && tok.skip(':')) {
```

Comment:
The scheme parsing API comments apply to this parsing function as well.

Reply:
Done.